Add CJS/ESM module exports

matteosacchetto opened the pull request:
With this pull request I am adding the support for both CJS and ESM module export. This will allow for example bundlers like webpack, rollup, vite, ... to threeshake any unnecessary code when writing web apps

To achieve this I ended up using [rollup](https://rollupjs.org/guide/en/) (easy to understand and configure) in combination with [esbuild](https://esbuild.github.io/) (for its speed)

I also update caniuse-lite to latest version (solved a warning during build) and typescript to latest version (due to an issue of version 4.4.2 (https://github.com/rollup/plugins/issues/983)

This pull request relates to discussion #49 
Resolves #79

GadQ commented on tsconfig.json:
```diff
@@ -1,9 +1,10 @@
-    "declaration": true,
-    "sourceMap": true
+    "moduleResolution": "node",
+    "target": "es2018",
```
I am no expert but shouldn't it be esnext, and then bundler could transpile it if necessary?

I recently saw an article about this:
https://github.com/frehner/modern-guide-to-packaging-js-library#target-modern-browsers

matteosacchetto replied:
Hi, thanks for the article! Interesting, I may be saving it for later.
**Quick disclaimer:** I am no expert either on this topic, I am still trying to figure it out 😅 

Anyway, so, regarding your query. It is true that setting it to `esnext` would allow those using a bundler to take advantage of the latest features and then transpile it to older versions, but this library is not targeted to browsers only, but also to node in general, and to project using simple JS, so without the use of TS.

For the ES version (`es2018`) I selected, I did so following this recommendations: https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping, aiming for node v10 as lowest.

Since in this project the code is tested on node12 as lowest node version, we could bump the config to `es2019` or even `es2020`, since now node 14 is the lowest LTS version of node. But I do not know if it is worth it to go to `esnext`, maybe there will be issues with older codebases...

But this is something I would let decide the maintainer of the library.

dolsem replied:
I also agree that setting to esnext would be ideal, this way we can have the TS compiler do as little transpilation as possible, making it the responsibility of the bundler, which is probably what we want.

matteosacchetto replied:
Ok then, I can change both lib and target it to `esnext` if you prefer it this way

dolsem replied:
sounds good, except I think lib should stay as it is, we were only talking about target

matteosacchetto replied:
Ok, got it!

dolsem commented:
Have you done testing to make sure that the resulting ESM target is tree shakable? Just because it's ESM-compatible doesn't guarantee that it's tree-shakable, so I just want to confirm. Honestly, I don't think bundling everything in a single index.js file is ideal, would be best to leave it as separate files, but as long as it's tree shakable and sourcemaps are generated correctly, it shouldn't matter much.

dolsem commented on rollup.config.js:
```diff
@@ -0,0 +1,85 @@
+    input: 'src/index.ts',
+    output: {
+      dir: 'dist/esm',
+      format: 'es',
```
shouldn't this be `esm`?

matteosacchetto replied:
From Rollup documentation
> ...
> `es` – Keep the bundle as an ES module file, suitable for other bundlers and inclusion as a <script type=module> tag in modern browsers (alias: esm, module)

`es` is the actual format name, while `esm` is an alias to `es`

For reference: https://rollupjs.org/guide/en/#outputformat

dolsem commented:
> I tried it using Vite (so Rollup + esbuild) and it was threeshaking perfectly. If you want to try it out wiht webpack, please feel free to do so. Easiest way to try this out is to clone the branch containing these changes, yarn pack to get the .tar.gz, then create a new webpack based project and install it from the .tar.gz archive, then see if it works with webpack too
>
> Regarding the single file vs multiple file, I opted for single file just to make the lib a bit smaller, but If you think multiple files is preferable then I will update it tomorrow toghether with the tsconfig options

Okay, thank you.

matteosacchetto commented:
I can confirm that also with webpack it is three shaking correctly! I've tested it on vue-cli project I have and these are the results

Code added
```js
import * as _ from "radash";
console.log(_.first([1, 2, 3]));
```

Here are the differences

- **Build size withouth above code**
  - chunk-vendors.45b3efd9.js - **269.15 KiB**
- **Build size with above code and current version of radash available on npm**
  - chunk-vendors.45b3efd9.js - **291.06 KiB**
- **Build size with above code and the new version of radash containing the changes in this PR**
  - chunk-vendors.45b3efd9.js - **269.26 KiB**

matteosacchetto commented:
`esnext` does not work with node 12. Following the mapping reported already previously (https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping) I will go to `es2019` to support node 12 and above

matteosacchetto commented:
No idea why the pipeline failed, it seems to be related to something  I did not actually touch (and btw, with same node version test pass locally...). If you can @rayepps, rerun the workflow.

Anyway, I added something with respect to the previous commit, which I will specify here:
```json
  "exports": {
    "import": "./dist/esm/index.mjs",
    "require": "./dist/cjs/index.cjs",
    "types": "./dist/index.d.ts"
  },
```

I added the exports section and I used the .cjs and .mjs extensions to make sure that node uses the correct loader.
It should be something minor, and three-shaking works, I can confirm that

gander commented:
bump

gander commented:
> No idea why the pipeline failed, it seems to be related to something I did not actually touch (and btw, with same node version test pass locally...). If you can @rayepps, rerun the workflow.
> 
> Anyway, I added something with respect to the previous commit, which I will specify here:
> 
> ```json
>   "exports": {
>     "import": "./dist/esm/index.mjs",
>     "require": "./dist/cjs/index.cjs",
>     "types": "./dist/index.d.ts"
>   },
> ```
> 
> I added the exports section and I used the .cjs and .mjs extensions to make sure that node uses the correct loader. It should be something minor, and three-shaking works, I can confirm that

It is dumb tests, based on time:
```
FAIL src/tests/async.test.ts (10.249 s)
  ● async module › _.sleep function › returns error when error is thrown

    assert.strictEqual(received, expected)

    Expected value to strictly be equal to:
      1665081140877
    Received:
      1665081140876

    Message:
      expected 1665081140876 to be at least 1665081140877

      216 |       await _.sleep(1000)
      217 |       const after = Date.now()
    > 218 |       assert.isAtLeast(after, before + 1000)
          |              ^
      21[9](https://github.com/rayepps/radash/actions/runs/3199576005/jobs/5225480375#step:5:10) |     })
      220 |   })
      221 |   

      at Object.<anonymous> (src/tests/async.test.ts:218:[14](https://github.com/rayepps/radash/actions/runs/3199576005/jobs/5225480375#step:5:15))```

rayepps commented:
@matteosacchetto I love this! I'm so excited to merge it. A bunch of PRs were merged recently so there is a new conflict but it's an easy one in the `yarn.lock`. I cloned your branch and ran it all locally, did some testing 👍 works perfectly.

@gander I would love to know how to improve that test cus I agree, it's dumb 😃

matteosacchetto commented:
@rayepps I have resolved the conflict with the yarn lock and kept typescript version 4.8.3
Since a couple of weeks back typescript 4.8.4 was release, do you want me to bump the version before merging?

rayepps commented:
@matteosacchetto can you run `rm yarn.lock && yarn` this way you'll just regenerate the yarn.lock (thats what I did locally)

I just published `8.0.0-rc.1` from this branch if you want to install it.

rayepps commented:
<img width="1241" alt="Screen Shot 2022-10-14 at 9 52 06 AM" src="https://user-images.githubusercontent.com/15269623/195900405-aae772d7-5f32-4817-bbda-dab776f3be4c.png">

This is awesome, I don't understand how/why but this CJS/ESM change brought the package size down by almost half, as well as providing the tree shaking 🤩

matteosacchetto commented:
> @matteosacchetto can you run `rm yarn.lock && yarn` this way you'll just regenerate the yarn.lock (thats what I did locally)
> 
> I just published `8.0.0-rc.1` from this branch if you want to install it.

I run `rm yarn.lock && yarn` and pushed the changes

matteosacchetto commented:
Again, issue on the same test mentioned above by @gander

matteosacchetto commented:
> <img alt="Screen Shot 2022-10-14 at 9 52 06 AM" width="1241" src="https://user-images.githubusercontent.com/15269623/195900405-aae772d7-5f32-4817-bbda-dab776f3be4c.png">
> 
> This is awesome, I don't understand how/why but this CJS/ESM change brought the package size down by almost half, as well as providing the tree shaking star_struck

The main reason of the smaller size is the bump on the tsconfig lib and target, so it is doing less transpilation and taking advantage of more modern JS

rayepps commented:
I made an issue to fix the issue in the tests (#102) 

cc: @gander

gustavopch commented:
I'm glad this was merged. When is it going to be released?

rayepps commented:
@gustavopch it will go out in 2 days (Monday) in v8.0.0. I'm working on a few other changes for the major release and I also happen to currently be traveling 😅
